File: mobile/android/air/SubModules/WalletCore/src/main/java/app/twallet/air/walletcore/moshi/explainedFee/MFee.kt
package app.twallet.air.walletcore.moshi.explainedFee

import com.squareup.moshi.JsonClass
import app.twallet.air.walletbasecontext.utils.smartDecimalsCount
import app.twallet.air.walletbasecontext.utils.toString
import app.twallet.air.walletcore.moshi.IApiToken
import java.math.BigInteger

@JsonClass(generateAdapter = true)
class MFee(
    var precision: MFeePrecision,
    val terms: MFeeTerms,
    /** The sum of `terms` measured in the native token */
    val nativeSum: BigInteger?,

    /** SwapOnly! Only the network fee terms (like `terms` but excluding our fee) */
    val networkTerms: MFeeTerms? = null
) {

    companion object {
        private const val ZERO_COUNT_SUBSCRIPT_MIN_COUNT = 6
    }

    val isNativeOnly: Boolean
        get() = (terms.token ?: BigInteger.ZERO) == BigInteger.ZERO
            && (terms.stars ?: BigInteger.ZERO) == BigInteger.ZERO

    fun toString(token: IApiToken, appendNonNative: Boolean): String {
        var result = ""
        val nativeToken = token.nativeToken

        networkTerms?.native?.takeIf { it > BigInteger.ZERO }?.let { native ->
            if (nativeToken != null) {
                result += native.toString(
                    nativeToken.decimals,
                    nativeToken.symbol,
                    native.smartDecimalsCount(nativeToken.decimals),
                    false,
                    zeroCountSubscriptMinCount = ZERO_COUNT_SUBSCRIPT_MIN_COUNT
                )
            }
        } ?: terms.native?.takeIf { it > BigInteger.ZERO }?.let { native ->
            if (nativeToken != null) {
                result += native.toString(
                    nativeToken.decimals,
                    nativeToken.symbol,
                    native.smartDecimalsCount(nativeToken.decimals),
                    false,
                    zeroCountSubscriptMinCount = ZERO_COUNT_SUBSCRIPT_MIN_COUNT
                )
            }
        }

        if (appendNonNative) {
            terms.token?.takeIf { it > BigInteger.ZERO }?.let { tokenAmount ->
                if (result.isNotEmpty()) {
                    result = " + $result"
                }
                result = tokenAmount.toString(
                    token.decimals,
                    token.symbol ?: "",
                    tokenAmount.smartDecimalsCount(token.decimals),
                    false,
                    zeroCountSubscriptMinCount = ZERO_COUNT_SUBSCRIPT_MIN_COUNT
                ) + result
            }

            terms.stars?.takeIf { it > BigInteger.ZERO }?.let { stars ->
                if (result.isNotEmpty()) {
                    result = " + $result"
                }
                result = stars.toString(
                    1,
                    "⭐️",
                    stars.smartDecimalsCount(1),
                    false
                ) + result
            }
        }

        if (result.isEmpty()) {
            result += BigInteger.ZERO.toString(
                0,
                nativeToken?.symbol ?: token.symbol ?: "",
                0,
                false
            )
        }

        return precision.prefix + result
    }
}

File: mobile/ios/Air/SubModules/WalletCore/Models/Fees/MFee.swift
//
//  MFee.swift
//  MyTonWalletAir
//
//  Created by Sina on 12/22/24.
//

import WalletContext
import WalletCoreTypes


public struct MFee: Equatable, Hashable, Codable, Sendable {
    
    public var precision: MFee.FeePrecision
    public var terms: MFee.FeeTerms
    /** The sum of `terms` measured in the native token */
    public var nativeSum: BigInt?
    
    public struct FeeTerms: Equatable, Hashable, Codable, Sendable {
        /** The fee part paid in the transferred token */
        let token: BigInt?

        /** The fee part paid in the chain's native token */
        let native: BigInt?
        
        public init(token: BigInt?, native: BigInt?) {
            self.token = token
            self.native = native
        }
    }

    public typealias FeePrecision = DecimalAmountFormatPrecision

    public init(precision: MFee.FeePrecision, terms: MFee.FeeTerms, nativeSum: BigInt?) {
        self.precision = precision
        self.terms = terms
        self.nativeSum = nativeSum
    }
    
    public var isNativeOnly: Bool {
        (terms.token ?? 0) == 0
    }
    
    public func toString(
        token: ApiToken,
        nativeToken: ApiToken
    ) -> String {
        var result = ""
        if let native = terms.native, native > 0 {
            let nativeAmount = TokenAmount(native, nativeToken)
            result += nativeAmount.formatted(.fee)
        }
        if let tokenAmount = terms.token, tokenAmount > 0 {
            if !result.isEmpty {
                result += " + "
            }
            let tokenAmount = TokenAmount(tokenAmount, token)
            result += tokenAmount.formatted(.fee)
        }
        if result.isEmpty {
            let zero = AnyDecimalAmount(0, decimals: 0, symbol: nativeToken.symbol, forceCurrencyToRight: true)
            result += zero.formatted(.none)
        }
        return precision.prefix + result
    }
}
